csharp/src/Apache.Arrow/Schema.cs

```diff
@@ -22,32 +22,31 @@ namespace Apache.Arrow
 {
     public partial class Schema
     {
-        public IReadOnlyDictionary<string, Field> Fields
-        {
-            get => _fieldsDictionary;
-        }
+        [Obsolete("Use `FieldList` instead")]
+        public IReadOnlyDictionary<string, Field> Fields => _fieldsDictionary;
```

Comment:
We could still support this API, couldn't we? We could use a custom `IReadOnlyDictionary` that stored multiple `Field` instances per string. And when someone tried accessing a field by name, if there was more than 1 field, we could define behavior of either:

* throw an exception
* return the last field with that name

Reply:
When enumerating the key value pairs, either:
- multiple key value pairs would have the same key
- some fields would be missing

In the remarks for the `IReadOnlyDictionary` docs, it is stated that:
> Each pair must have a unique key.

This means the former would violate this and could lead to behaviour that the user would not expect.

If both 
- a single field is returned (instead of throwing an exception) when someone tries accessing a field by name for which there are multiple fields
- duplicate fields are skipped when enumerating

then this is equivalent to just skipping duplicate keys when building the dictionary.

This API can still be supported by skipping duplicate keys when building the dictionary, but that would lead to the dictionary being an inaccurate representation of the state of the schema when there are multiple fields with the same name. Because of this, deprecating it could make it clear that using it is to be avoided.

Reply:
Do we want to add an `ILookup<string, Field>` which might be handy on wide schemas?  It's still an API break but might be useful.

Reply:
@westonpace Exposing an `ILookup<string, Field>` alongside an `IReadOnlyList<Field>` would provide more flexibility to users. I think doing so could be a good idea

Reply:
I've added the `ILookup<string, Field>`